File: Sean的leetcode/Leetcode/Smallest Rectangle Enclosing Black Pixels.cpp
//Binary Search O(mlogn + nlogm)
//Search for the boundary, find the first line that contains one
class Solution {
public:
    vector<vector<char>> img;
    int minArea(vector<vector<char>>& image, int x, int y) {
        img = image;
        int rows = image.size();
        int cols = rows? image[0].size() : 0;
        if (!rows || !cols) return 0;
        int up = binarysearch(0, x, 0, cols-1, 0);
        int down = binarysearch(x, rows-1, 0, cols-1, 1);
        int left = binarysearch(0, y, up, down, 2);
        int right = binarysearch(y, cols-1, up, down, 3);
        return (down - up + 1) * (right - left + 1);
    }
    int binarysearch(int l, int r, int low, int high, int dir) {
        
        if (dir == 0) {
            while (l + 1 < r) {
                int mid = l + (r - l) / 2;
                int j;
                for (j = low; j <= high && img[mid][j] != '1'; j++);
                if (j <= high) {
                    r = mid;
                }
                else {
                    l = mid;
                }
            }
            int j;
            for (j = low; j <= high && img[l][j] != '1'; j++);
            return j <= high? l : r;
        }
        else if (dir == 1) {
            while (l + 1 < r) {
                int mid = l + (r - l) / 2;
                int j;
                for (j = low; j <= high && img[mid][j] != '1'; j++);
                if (j <= high) {
                    l = mid;
                }
                else {
                    r = mid;
                }
            }
            int j;
            for (j = low; j <= high && img[r][j] != '1'; j++);
            return j <= high? r : l;
        }
        else if (dir == 2) {
            while (l + 1 < r) {
                int mid = l + (r - l) / 2;
                int i;
                for (i = low; i <= high && img[i][mid] != '1'; i++);
                if (i <= high) {
                    r = mid;
                }
                else {
                    l = mid;
                }
            }
            int i;
            for (i = low; i <= high && img[i][l] != '1'; i++);
            return i <= high? l : r;
        }
        else {
            while (l + 1 < r) {
                int mid = l + (r - l) / 2;
                int i;
                for (i = low; i <= high && img[i][mid] != '1'; i++);
                if (i <= high) {
                    l = mid;
                }
                else {
                    r = mid;
                }
            }
            int i;
            for (i = low; i <= high && img[i][r] != '1'; i++);
            return i <= high? r : l;
        }
    }
};
//DFS, Time: O(mn)
class Solution {
public:
    void dfs(vector<vector<char>>& image, int row, int col, int &min_row, int &max_row, int &min_col, int &max_col)
    {
        int dir[] = {0, 1, 0, -1, 0};
        int M_row = image.size();

        if (!M_row) return;

        int M_col = image[0].size();

        if (row < 0 || col < 0 || row >= M_row || col >= M_col)
            return;

        if (image[row][col] == '1')
        {
            if (row < min_row) min_row = row;
            if (row > max_row) max_row = row;
            if (col < min_col) min_col = col;
            if (col > max_col) max_col = col;
            image[row][col] = 0;
            for (int i = 0; i < 4; i++)
                dfs(image, row+dir[i], col+dir[i+1], min_row, max_row, min_col, max_col);
        }
    }

    int minArea(vector<vector<char>>& image, int x, int y) {
        int min_row = INT_MAX, min_col = INT_MAX;
        int max_row = INT_MIN, max_col = INT_MIN;

        int M_row = image.size();

        if (!M_row) return 0;

        dfs(image, x, y, min_row, max_row, min_col, max_col);

        return (max_row - min_row + 1) * (max_col - min_col + 1);
    }
};

File: Sean的leetcode/python/Smallest Rectangle Enclosing Black Pixels.py
###
# Binary Search
# Time Complexity: O(nlogm+mlogn)
# Space Complexity: O(1)
###
class Solution(object):
    def minArea(self, image, x, y):
        """
        :type image: List[List[str]]
        :type x: int
        :type y: int
        :rtype: int
        """
        rows = len(image)
        cols = len(image[0])
        if not rows or not cols:
            return 0
        up = self.binary_search(image, 0, x, 0, cols-1, 'u')
        down = self.binary_search(image, x, rows-1, 0, cols-1, 'd')
        left = self.binary_search(image, up, down, 0, y, 'l')
        right = self.binary_search(image, up, down, y, cols-1, 'r')
        return ( down - up + 1) * (right - left + 1)
    
    def binary_search(self, image, si, ei, sj, ej, drc):
        if drc == 'u':
            l = si
            r = ei
            while l + 1 < r:
                mid = (l + r)/2
                if self.has_black(image, mid, mid, sj, ej):
                    r = mid
                else:
                    l = mid
            return l if self.has_black(image, l, l, sj, ej) else r
        elif drc == 'd':
            l = si
            r = ei
            while l + 1 < r:
                mid = (l + r)/2
                if self.has_black(image, mid, mid, sj, ej):
                    l = mid
                else:
                    r = mid      
            return r if self.has_black(image, r, r, sj, ej) else l
        elif drc == 'l':
            l = sj
            r = ej
            while l + 1 < r:
                mid = (l + r)/2
                if self.has_black(image, si, ei, mid, mid):
                    r = mid
                else:
                    l = mid
            return l if self.has_black(image, si, ei, l, l) else r
        else:
            l = sj
            r = ej
            while l + 1 < r:
                mid = (l + r)/2
                if self.has_black(image, si, ei, mid, mid):
                    l = mid
                else:
                    r = mid
            return r if self.has_black(image, si, ei, r, r) else l
        
    def has_black(self, image, si, ei, sj, ej):  
        return any(image[i][j]=="1" for i in xrange(si, ei+1) for j in xrange(sj, ej+1))
        
                                        
                            
                    